include/pybind11/cast.h

```diff
@@ -907,6 +907,12 @@ struct handle_type_name<kwargs> {
 
 template <typename type>
 struct pyobject_caster {
+    template <typename T = type, enable_if_t<std::is_same<T, handle>::value, int> = 0>
+    pyobject_caster() : value() {}
+
+    template <typename T = type, enable_if_t<std::is_base_of<object, T>::value, int> = 0>
```

Comment:
I am not sure about the changes on the caster: So without a default ctor, this ctor is needed on pyobject_caster? At the very least, we should have a comment about why there are necessary. I am also worried that there are edge cases where this might break when it is used in downstream applications.

Thoughts @rwgk @henryiii

Reply:
IIUC we have a choice:

* Either the change here.
* Or we need to make `anyset` and `frozenset` default constructible.

> https://github.com/pybind/pybind11/pull/3901#issuecomment-1114261772

I agree but think it is not just "still safe" but actually safer: if the previously default-constructed object is accidentally used the symptoms could be highly confusing (silent failures). A noisy failure (segfault b/o nullptr) seems better.

I'll run our (Google's) global testing with this PR to see if there are breakages and report back here. I might have the results only tomorrow.

Reply:
No breakages! This PR is great. If anyone was previously relying on the default-constructed `value`, they will probably be glad to see that exposed as an obvious bug now.